### DISCUSSION TITLE
Start designing an official or community-driven vcpkg dependency provider implementation

### DISCUSSION THREAD

**Discussion by theblackunknown** [microsoft/vcpkg#37664] [Ideas]
This discussion is about starting implementing a CMake dependency provider for `vcpkg` the proper way.

Based on search for similar discussions, there are a couple of past/pending efforts but they seem to be on hold given the last update date:
- microsoft/vcpkg#31062
- https://github.com/microsoft/vcpkg/pull/27950#issuecomment-1507406288

Ideally I would like to start making an "official" implementation for this but based on https://github.com/microsoft/vcpkg/pull/27950#issuecomment-1507406288 maybe this is not desirable and we would provide a community driven one, the same way we do for triplet useful for the community but not tested on the CI ?

I do believe we can start one without breaking backward compatibility with the current `scripts/buildsystems/vcpkg.cmake`, for example we can create another `scripts/buildsystems(/community)/vcpkg-dependency-provider.cmake` which share common code with the former.

**Comment by Neumann-A** [comment]
The question rather is:
What would be the benefit of it? cmake dependency providers are just an upstream blessed overwrite of `find_package`

**Reply by theblackunknown** [reply]
In my opinion, there may not be so much benefit from a `vcpkg` point of view, but from a generic CMake project point of view it allows for a consistent customization point based on something blessed by CMake.

Concretely we would have a consistent approach to bring dependency in a project, whether we want to use `vcpkg`, `conan`, an homemade solution or anything else which will be available in the future.

**Reply by Neumann-A** [reply]
> Concretely we would have a consistent approach to bring dependency in a project

You don't need dependency providers to use `CMAKE_TOP_LEVEL_INCLUDE`. You can already use it with `vcpkg.cmake`. But be aware that this has implications for `try_compile` since `CMAKE_TOOLCHAIN_FILE` is passed on but `CMAKE_TOP_LEVEL_INCLUDE` is not

**Reply by theblackunknown** [reply]
I thought that your PR microsoft/vcpkg#31062 showed the minimal required change to make `vcpkg.cmake` works as a dependency provider ?

The current implementation overrides `find_package` implementation and does not call `cmake_language(SET_DEPENDENCY_PROVIDER ...)` so it does not look like this can work, am I missing something ?

**Reply by Neumann-A** [reply]
> I thought that your PR https://github.com/microsoft/vcpkg/pull/31062 showed the minimal required change to make vcpkg.cmake works as a dependency provider ?

The problem is you will still need to have the `find_package` override since vcpkg-cmake-wrapper.cmake requires it. So for versioning and backward compat it is always needed

**Reply by theblackunknown** [reply]
Good point ! I forgot about those `vcpkg-cmake-wrapper.cmake`, but maybe this is something a dependency provider could implement as well as part of its `find_package` override ?
It would make sense in my opinion as using `find_package(BYPASS_PROVIDER ...)` will then rightfully skip those too.

**Reply by dg0yt** [reply]
Find something better than `vcpkg-cmake-wrapper.cmake` could be desired: Wrappers don't see the normal `find_package` interface variables.
